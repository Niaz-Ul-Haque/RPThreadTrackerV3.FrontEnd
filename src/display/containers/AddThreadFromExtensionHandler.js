// #region imports
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Container, Row, Col, Card, CardBlock } from 'reactstrap';
import PropTypes from 'prop-types';

import { fetchCharacters, fetchUser, openUpsertThreadModal } from '../../infrastructure/actions';
import { getCharactersSortedByIdentifier } from '../../infrastructure/selectors';

import LoadingIndicator from '../shared/LoadingIndicator';
import ModalContainer from '../shared/modals/ModalContainer';

import withPageViewTracker from '../../infrastructure/withPageViewTracker';

import { getThreadDataFromExtensionQuery } from '../../utility';
// #endregion imports

const propTypes = {
	sortedCharacters: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	fetchCharacters: PropTypes.func.isRequired,
	fetchUser: PropTypes.func.isRequired,
	isUpsertThreadModalOpen: PropTypes.bool.isRequired,
	openUpsertThreadModal: PropTypes.func.isRequired,
	user: PropTypes.shape({
		id: PropTypes.string
	}).isRequired
};
const mapStateToProps = (state) => {
	const {
		user,
		ui
	} = state;
	const sortedCharacters = getCharactersSortedByIdentifier(state);
	return {
		user,
		sortedCharacters,
		isUpsertThreadModalOpen: ui.isUpsertThreadModalOpen
	};
};
class AddThreadFromExtensionHandler extends Component {
	constructor() {
		super();
		this.state = { hasOpenedModal: false };
		this.shouldOpenModal = this.shouldOpenModal.bind(this);
	}
	componentDidMount() {
		this.props.fetchUser();
		this.props.fetchCharacters();
	}
	componentWillReceiveProps(nextProps) {
		if (this.shouldOpenModal(nextProps.user, nextProps.sortedCharacters)) {
			const thread = getThreadDataFromExtensionQuery(nextProps.sortedCharacters);
			this.props.openUpsertThreadModal(thread);
		}
	}
	shouldOpenModal(user, characters) {
		if (!user || !user.id) {
			return false;
		}
		if (!characters || !characters.length) {
			return false;
		}
		if (this.state.hasOpenedModal) {
			return false;
		}
		this.setState({ hasOpenedModal: true });
		return true;
	}
	showLoadingIndicator() {
		return (
			<LoadingIndicator
				data-spec="extension-handler-loader"
				style={{
					width: 50,
					height: 50,
					position: 'absolute',
					top: '50%',
					left: '50%',
					transform: 'translate(-50%, -50%)'
				}}
			/>
		);
	}
	showLayout() {
		return (
			<div className="app flex-row align-items-center" data-spec="layout-app">
				<ModalContainer />
				{this.state.hasOpenedModal && !this.props.isUpsertThreadModalOpen &&
					<Container data-spec="extension-handler-success-message">
						<Row className="justify-content-center">
							<Col md="6">
								<Card className="login-box p-4">
									<CardBlock className="card-body text-center">
										<p>
											You can now close this window.
										</p>
									</CardBlock>
								</Card>
							</Col>
						</Row>
					</Container>
				}
			</div>
		);
	}
	render() {
		if (!this.props.user.id || !this.props.sortedCharacters.length) {
			return this.showLoadingIndicator();
		}
		return this.showLayout();
	}
}

AddThreadFromExtensionHandler.propTypes = propTypes;
export default connect(mapStateToProps, {
	fetchCharacters,
	fetchUser,
	openUpsertThreadModal
})(withPageViewTracker(AddThreadFromExtensionHandler));
