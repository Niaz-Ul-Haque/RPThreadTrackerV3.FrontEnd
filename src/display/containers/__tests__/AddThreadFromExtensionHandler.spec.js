// #region imports
import React from 'react';
import { getSpecWrapper, shallowWithState } from '../../../../config/tests/helpers.unit';
import AddThreadFromExtensionHandler from '../AddThreadFromExtensionHandler';
// #endregion imports

// #region mocks
jest.mock('../../../infrastructure/actions', () => ({}));
jest.mock('../../../infrastructure/selectors', () => ({
	getCharactersSortedByIdentifier: state => state.sortedCharacters
}));
jest.mock('../../shared/styled/Card', () => 'Card');
jest.mock('../../shared/modals/ModalContainer', () => 'ModalContainer');
jest.mock('../../shared/loading/LoadingIndicator', () => 'LoadingIndicator');
jest.mock('../../../infrastructure/withPageViewTracker', () => Component => Component);
jest.mock('../../../utility', () => ({
	getThreadDataFromExtensionQuery: () => ({ characterId: 3, postId: '13579' })
}));
// #endregion mocks

const createTestProps = propOverrides => ({
	fetchCharacters: jest.fn(),
	fetchUser: jest.fn(),
	openUpsertThreadModal: jest.fn(),
	...propOverrides
});

const createTestState = stateOverrides => ({
	sortedCharacters: [],
	user: {},
	ui: { isUpsertThreadModalOpen: false },
	...stateOverrides
});

const createTestStateWithUser = stateOverrides => createTestState({
	user: { id: '12345' },
	...stateOverrides
});

const createTestStateWithUserAndCharacters = stateOverrides => createTestState({
	user: { id: '12345' },
	sortedCharacters: [{ characterName: 'A Character', characterId: 1 }, { characterName: 'B Character', characterId: 2 }],
	...stateOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot if user is not loaded', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot if characters are not loaded', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestStateWithUser();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot if user and characters are loaded and modal open', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestStateWithUserAndCharacters({ ui: { isUpsertThreadModalOpen: true } });
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot if user and characters are loaded and modal has been closed', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestStateWithUserAndCharacters({ ui: { isUpsertThreadModalOpen: true } });
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.setProps(createTestStateWithUserAndCharacters({ isUpsertThreadModalOpen: false }));
			expect(element).toMatchSnapshot();
		});
	});
	describe('loading indicator', () => {
		it('should be displayed if user not loaded', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(getSpecWrapper(element, 'extension-handler-loader')).toHaveLength(1);
			expect(getSpecWrapper(element, 'layout-app')).toHaveLength(0);
		});
		it('should be displayed if characters not loaded', () => {
			const props = createTestProps();
			const state = createTestStateWithUser();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(getSpecWrapper(element, 'extension-handler-loader')).toHaveLength(1);
			expect(getSpecWrapper(element, 'layout-app')).toHaveLength(0);
		});
		it('should not be displayed if user and characters loaded', () => {
			const props = createTestProps();
			const state = createTestStateWithUserAndCharacters();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(getSpecWrapper(element, 'extension-handler-loader')).toHaveLength(0);
			expect(getSpecWrapper(element, 'layout-app')).toHaveLength(1);
		});
	});
	describe('success message', () => {
		it('should be hidden if user not loaded', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(getSpecWrapper(element, 'extension-handler-success-message')).toHaveLength(0);
			expect(getSpecWrapper(element, 'layout-app')).toHaveLength(0);
		});
		it('should be hidden if characters not loaded', () => {
			const props = createTestProps();
			const state = createTestStateWithUser();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(getSpecWrapper(element, 'extension-handler-success-message')).toHaveLength(0);
			expect(getSpecWrapper(element, 'layout-app')).toHaveLength(0);
		});
		it('should be hidden if user and characters loaded', () => {
			const props = createTestProps();
			const state = createTestStateWithUserAndCharacters();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(getSpecWrapper(element, 'extension-handler-success-message')).toHaveLength(0);
			expect(getSpecWrapper(element, 'layout-app')).toHaveLength(1);
		});
	});
});

describe('behavior', () => {
	describe('componentDidMount', () => {
		it('should retrieve user', () => {
			const fetchUser = jest.fn();
			const props = createTestProps({ fetchUser });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchUser).toHaveBeenCalledTimes(1);
		});
		it('should retrieve characters', () => {
			const fetchCharacters = jest.fn();
			const props = createTestProps({ fetchCharacters });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchCharacters).toHaveBeenCalledTimes(1);
		});
	});
	describe('componentWillReceiveProps', () => {
		it('should not trigger modal open if user is not loaded', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.setProps({ user: {} });
			expect(openUpsertThreadModal).toHaveBeenCalledTimes(0);
		});
		it('should not trigger modal open if characters are not loaded', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.setProps(createTestStateWithUser());
			expect(openUpsertThreadModal).toHaveBeenCalledTimes(0);
		});
		it('should trigger modal open if user and characters are loaded', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.setProps(createTestStateWithUserAndCharacters());
			expect(openUpsertThreadModal).toHaveBeenCalledTimes(1);
		});
		it('should not display success message if user is not loaded', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.setProps({ user: {} });
			const successMessage = getSpecWrapper(element, 'extension-handler-success-message');
			expect(successMessage).not.toExist();
		});
		it('should not display success message if characters are not loaded', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.setProps(createTestStateWithUser());
			const successMessage = getSpecWrapper(element, 'extension-handler-success-message');
			expect(successMessage).not.toExist();
		});
		it('should not display success message if user and characters are loaded and modal open', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.setProps(createTestStateWithUserAndCharacters({ isUpsertThreadModalOpen: true }));
			const successMessage = getSpecWrapper(element, 'extension-handler-success-message');
			expect(successMessage).not.toExist();
		});
		it('should display success message if user and characters are loaded and modal has been closed', () => {
			const openUpsertThreadModal = jest.fn();
			const props = createTestProps({ openUpsertThreadModal });
			const state = createTestState();
			const jsx = (<AddThreadFromExtensionHandler {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.setProps(createTestStateWithUserAndCharacters({ isUpsertThreadModalOpen: true }));
			element.setProps(createTestStateWithUserAndCharacters({ isUpsertThreadModalOpen: false }));
			const successMessage = getSpecWrapper(element, 'extension-handler-success-message');
			expect(successMessage).toExist();
		});
	});
});
