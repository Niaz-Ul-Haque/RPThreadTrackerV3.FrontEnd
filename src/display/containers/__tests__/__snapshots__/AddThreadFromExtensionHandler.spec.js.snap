// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot if characters are not loaded 1`] = `
<LoadingIndicator
  data-spec="extension-handler-loader"
  style={
    Object {
      "height": 50,
      "left": "50%",
      "position": "absolute",
      "top": "50%",
      "transform": "translate(-50%, -50%)",
      "width": 50,
    }
  }
/>
`;

exports[`rendering snapshots should render valid snapshot if user and characters are loaded and modal has been closed 1`] = `
<div
  className="app flex-row align-items-center"
  data-spec="layout-app"
>
  <ModalContainer />
  <Container
    data-spec="extension-handler-success-message"
    tag="div"
  >
    <Row
      className="justify-content-center"
      tag="div"
    >
      <Col
        md="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <Styled(Card)
          className="login-box p-4"
        >
          <CardBody
            className="card-body text-center"
            tag="div"
          >
            <p>
              You can now close this window.
            </p>
          </CardBody>
        </Styled(Card)>
      </Col>
    </Row>
  </Container>
</div>
`;

exports[`rendering snapshots should render valid snapshot if user and characters are loaded and modal open 1`] = `
<div
  className="app flex-row align-items-center"
  data-spec="layout-app"
>
  <ModalContainer />
</div>
`;

exports[`rendering snapshots should render valid snapshot if user is not loaded 1`] = `
<LoadingIndicator
  data-spec="extension-handler-loader"
  style={
    Object {
      "height": 50,
      "left": "50%",
      "position": "absolute",
      "top": "50%",
      "transform": "translate(-50%, -50%)",
      "width": 50,
    }
  }
/>
`;
